### PR TITLE
RangeControl: Add support for large 40px number input size

### PR DESCRIPTION
**Pull request by richtabor**
## What?

Add support for the `size="__unstable-large"` prop within the RangeControl component, to support the effort towards consistent components across the board re https://github.com/WordPress/gutenberg/issues/46734. 

Related to https://github.com/WordPress/gutenberg/pull/49101 and https://github.com/WordPress/gutenberg/issues/46741.

## Why?
Support the consistent component effort in https://github.com/WordPress/gutenberg/issues/46734.

## How?
1. Add a size prop. 
2. Tweak styling to align:center from start (I don't think this changes the component's layout for default sizes)
3. Zero out margin bottom, as it was not always zeroed—like when I tested in https://github.com/WordPress/gutenberg/pull/49101.
4. Set width of large input to `size(20)`.
5. Add `size="__unstable-large"` to the Cover block's Overlay Opacity control for testing. 

## Testing Instructions

1. Open a Post or Page.
6. Insert a Cover block.
7. Add an image or pick a color.
8. Open Block Inspector.
9. Select the "Styles" tab.
10. See the larger 40px input on the "Overlay Opacity" RangeControl.

## Screenshots or screencast 

Before: 

CleanShot 2023-03-15 at 11 44 24

After: 

CleanShot 2023-03-15 at 12 42 23

### CODE REVIEW

**Comment by jasmussen** [comment]
This makes me happy. @mirka or @ciampo, what do you think?

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (mirka) [inline packages/components/src/range-control/styles/range-control-styles.ts]: Why is this necessary? (Both the `margin-bottom` and the `!important`)

File: packages/components/src/range-control/index.tsx
Comment (mirka) [inline packages/components/src/range-control/index.tsx]: For the RangeControl interface, we'll want to go straight to the `__next40pxDefaultSize` API as outlined in #46741. (Better not introduce a new `size` prop that will be obsolete in a few months!)

```suggestion
		__next40pxDefaultSize = false,
```

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (mirka) [inline packages/components/src/range-control/styles/range-control-styles.ts]: Are these `!important`s necessary?

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (mirka) [inline packages/components/src/range-control/styles/range-control-styles.ts:1]: This also needs to support the `withInputField={ false }` [case](https://wordpress.github.io/gutenberg/?path=/story/components-rangecontrol--default&args=help:;__nextHasNoMarginBottom:true;withInputField:false&globals=marginChecker:show). Meaning, the slider has to be positioned correctly within a 30px/40px height regardless of whether the component is rendering an input field, so it's aligned with any other controls rendered beside it.

File: packages/components/src/range-control/index.tsx
Comment (richtabor) [reply packages/components/src/range-control/index.tsx]: `__next40pxDefaultSize` isn't available yet, is it?

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (richtabor) [reply packages/components/src/range-control/styles/range-control-styles.ts]: They were already in place; I can check if we really need them though.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (richtabor) [reply packages/components/src/range-control/styles/range-control-styles.ts]: The margin-bottom does not get always zero'ed out, resulting in the track/thumb not being aligned properly. It shouldn't have margin applied to the input within the control.

Comment (mirka) [reply packages/components/src/range-control/styles/range-control-styles.ts]: I'm guessing you're referring to the margin-bottom [added](https://github.com/WordPress/gutenberg/blob/584c1d27a92db85491fe4bf3eb883d689faa3f5f/packages/block-editor/src/components/block-inspector/style.scss#L13-L19) by the `block-inspector` stylesheet? This is a `block-editor` package concern, so it should be dealt with in that package rather than in `@wordpress/components`.

Can it be addressed similarly to the FocalPointPicker here?

https://github.com/WordPress/gutenberg/blob/584c1d27a92db85491fe4bf3eb883d689faa3f5f/packages/block-editor/src/components/block-inspector/style.scss#L21-L25

File: packages/components/src/range-control/index.tsx
Comment (mirka) [reply packages/components/src/range-control/index.tsx]: Yes, this will be the first occurrence of a `__next40pxDefaultSize` prop in the codebase. Inside of the RangeControl, NumberControl will still need to receive a `size` prop. So basically something like:

```jsx
					<InputNumber
						size={ __next40pxDefaultSize ? '__unstable-large' : 'default' }
```

**Comment by ciampo** [comment]
Hey @richtabor , just checking if you need any help with this PR or if you put it on hold for any specific reasons?

**Comment by richtabor** [comment]
> Hey @richtabor , just checking if you need any help with this PR or if you put it on hold for any specific reasons?

Sorry, just got caught up on other fronts. Feel free to take over if you'd like.

**Comment by chad1008** [comment]
Picked this PR up today, I'll wrap up the rest of the necessary changes 👍

**Comment by richtabor** [comment]
> Picked this PR up today, I'll wrap up the rest of the necessary changes 

Let us know when it's ready for a new review. 👍

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts:1]: Good eye! The vertical placement was being handled by flex alignment on the `Root` subcomponent, so I've applied a conditional `min-height` there. When the `__next40present`, it will set the height of the that subcomponent to `40px`, otherwise it will fall back to the previous default height of `30px`.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: These were put into place[ to fix a visual regression a couple of years ago](https://github.com/WordPress/gutenberg/pull/28816). While we haven't yet completed the long term fix recommended at the time of refactoring `RangeControl` to use `HStack`, it does look like other more recent changes have alleviated the original issue. Without these `!importants` I'm not able to replicate the original problem shows in that PR, so I've removed them for now.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: This worked well, thanks for the suggestion @mirka

File: packages/components/src/range-control/index.tsx
Comment (chad1008) [reply packages/components/src/range-control/index.tsx]: This should be in place now, but let me know how it looks!

**Comment by chad1008** [comment]
I've taken a pass at all outstanding comments - let me know if there's any new or updated feedback!

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md]: Formatting nit: there should be an empty line after the section header, and no empty line between each list item

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [inline packages/components/src/range-control/styles/range-control-styles.ts]: We can simplify the logic in this function a bit:

```suggestion
const inputNumberWidth = ( { size }: InputNumberProps ) => {
	const sizes = {
		default: space( 16 ),
		'__unstable-large': space( 20 ),
	};

	return css`
		width: ${ sizes[ size ?? 'default' ] };
	`;
};
```

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [inline packages/components/src/range-control/styles/range-control-styles.ts]: Out of curiosity, is there a design spect about the number input's width changing too ?

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [inline packages/components/src/range-control/styles/range-control-styles.ts]: Are these `min-height` styles necessary? I guess if the root wrapper's height would be set correctly anyway, since it's "pushed" by the height of the number input.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: Good call

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: This was my solution to @mirka's [point](https://github.com/WordPress/gutenberg/pull/49105#discussion_r1138612084) that the component needs to support the `withInputField={ false }` case. When we have an input you're right, the wrapper's height will be pushed to the correct height automatically. When we render without an input though, we need to somehow enforce that height on the wrapper directly.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: 🤔 I'm not sure. I'll look around, but ccing @richtabor in case you recall. This change predates my work on the PR.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [reply packages/components/src/range-control/styles/range-control-styles.ts]: Gotcha.

Seeing that we have a `rangeHeightValue` variable used across the component, a neater approach could be to provide another `rangeHeightValueNext` variable (set to `40`), and throughout the component's styles switch between the two variables based on the `__next40pxDefaultSize` prop.

But I also understand that it would be more work. Given that the `__next40pxDefaultSize` is going to be around only temporarily, we can probably keep styles as they are currently, and clean them up again once we remove the `__next40pxDefaultSize` prop.

Maybe we can just clean it up a little by moving all the min-height assignment inside the same function, and adding a TODO comment:

```diff
diff --git a/packages/components/src/range-control/styles/range-control-styles.ts b/packages/components/src/range-control/styles/range-control-styles.ts
index e0da254e12..d28d4f1646 100644
--- a/packages/components/src/range-control/styles/range-control-styles.ts
+++ b/packages/components/src/range-control/styles/range-control-styles.ts
@@ -23,16 +23,20 @@ import type {
 } from '../types';
 
 const rangeHeightValue = 30;
+const rangeHeightValueNext = 40;
 const railHeight = 4;
 const rangeHeight = () =>
 	css( { height: rangeHeightValue, minHeight: rangeHeightValue } );
 const thumbSize = 12;
 
-const deprecatedDefaultHeight = ( {
+const rootMinHeight = ( {
 	__next40pxDefaultSize,
-}: Pick< RangeControlProps, '__next40pxDefaultSize' > ) => {
-	return ! __next40pxDefaultSize && css( { minHeight: rangeHeightValue } );
-};
+}: Pick< RangeControlProps, '__next40pxDefaultSize' > ) =>
+	css( {
+		minHeight: __next40pxDefaultSize
+			? rangeHeightValueNext
+			: rangeHeightValue,
+	} );
 
 type RootProps = Pick< RangeControlProps, '__next40pxDefaultSize' >;
 export const Root = styled.div< RootProps >`
@@ -44,9 +48,9 @@ export const Root = styled.div< RootProps >`
 	position: relative;
 	touch-action: none;
 	width: 100%;
-	min-height: 40px;
 
-	${ deprecatedDefaultHeight }
+	/* TODO: remove after removing the __next40pxDefaultSize prop */
+	${ rootMinHeight }
 `;
 
 const wrapperColor = ( { color = COLORS.ui.borderFocus }: WrapperProps ) =>


```

File: packages/components/CHANGELOG.md
Comment (ciampo) [reply packages/components/CHANGELOG.md]: The conversation is marked as resolved but without any changes to the code — maybe you forgot to push the commit?

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [reply packages/components/src/range-control/styles/range-control-styles.ts]: Maybe @jasmussen can give us a 👍  or a 👎  , since rich may be busy on other fronts

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [reply packages/components/src/range-control/styles/range-control-styles.ts]: We can actually simplify this a little bit further, by re-using `NumberControl` prop types

```diff
diff --git a/packages/components/src/range-control/styles/range-control-styles.ts b/packages/components/src/range-control/styles/range-control-styles.ts
index e0da254e12..a1ed8852bf 100644
--- a/packages/components/src/range-control/styles/range-control-styles.ts
+++ b/packages/components/src/range-control/styles/range-control-styles.ts
@@ -10,9 +10,8 @@ import styled from '@emotion/styled';
 import NumberControl from '../../number-control';
 import { COLORS, reduceMotion, rtl } from '../../utils';
 import { space } from '../../ui/utils/space';
-
+import type { NumberControlProps } from '../../number-control/types';
 import type {
-	InputNumberProps,
 	RangeMarkProps,
 	RailProps,
 	ThumbProps,
@@ -302,20 +301,17 @@ export const Tooltip = styled.span< TooltipProps >`
 	) }
 `;
 
-const inputNumberWidth = ( { size }: InputNumberProps ) => {
+const inputNumberWidth = ( {
+	size = 'default',
+}: Pick< NumberControlProps, 'size' > ) => {
 	const sizes = {
+		small: space( 16 ),
 		default: space( 16 ),
 		'__unstable-large': space( 20 ),
 	};
 
-	if ( size === '__unstable-large' ) {
-		return css( {
-			width: `${ sizes[ '__unstable-large' ] }`,
-		} );
-	}
-
 	return css`
-		width: ${ sizes.default };
+		width: ${ sizes[ size ?? 'default' ] };
 	`;
 };
 
diff --git a/packages/components/src/range-control/types.ts b/packages/components/src/range-control/types.ts
index ffccd327e3..4f7c5571fd 100644
--- a/packages/components/src/range-control/types.ts
+++ b/packages/components/src/range-control/types.ts
@@ -277,12 +277,6 @@ export type TrackProps = {
 	trackColor: CSSProperties[ 'color' ];
 };
 
-export type Size = 'default' | '__unstable-large';
-
-export type InputNumberProps = {
-	size?: Size;
-};
-
 export type UseControlledRangeValueArgs = {
 	/**
 	 * The initial value.

```

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (jasmussen) [reply packages/components/src/range-control/styles/range-control-styles.ts]: I could be missing context, so I'll defer to Rich in case he chimes in.

But I don't recall anywhere we're using 20px for spacing items. We use 8px and 16px, in some places 12px, never 10 or 20.

Comment (ciampo) [reply packages/components/src/range-control/styles/range-control-styles.ts]: `space( 20 )` means `20 * 4px = 80px.

Prior to this PR, the input field was always `space( 16 )` (ie. `64px`) wide. Rich added a change that would make the input's width change to `80px` when its height is `40px`.

@jasmussen , we were wondering if it's ok for the input field to be wider when its height is `40px`, or if it should stay the same width as previously

Comment (jasmussen) [reply packages/components/src/range-control/styles/range-control-styles.ts]: Ah, thanks for the clarification. In this case, anything goes, because I think for the particular opacity control in question, it's actually wrong as is, and should follow the pattern of the other more modern slides, a la radius:

Screenshot 2023-06-07 at 16 02 03

But that's fine to do separately, there could be a valid discussion around that, so nothing to block work here.

**Comment by chad1008** [comment]
Okay, I've addressed the latest pass of feedback (sorry for the delayed turnaround). I think that should address everything, unless we want to do away with the increased width discussed here. Based on @jasmussen's answers (thanks for those!) it looks like we could go either way on that for now.

**Comment by jasmussen** [comment]
I'll defer to Rich, but happy to provide a green check when things settle!

File: packages/components/CHANGELOG.md
Comment (chad1008) [reply packages/components/CHANGELOG.md]: Sorry about that,  I think I forgot to push some actual changes I'd made. Should all be on the PR now.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (mirka) [inline packages/components/src/range-control/styles/range-control-styles.ts]: We can remove this because the `small` case will never happen.

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (mirka) [inline packages/components/src/range-control/styles/range-control-styles.ts]: Just a stylistic thing, but as a courtesy for when we actually have to remove the temporary `__next*` props, I find it cleaner to structure the code so that the `! __next` case is already handled as the exception case.

For example:

```suggestion
	min-height: 40px;
	
	${ deprecatedHeight };
```

where `deprecatedHeight` only handles the `! __next40pxDefaultSize` case. The clean up will be easier this way. ([Actual example](https://github.com/WordPress/gutenberg/blob/33fbd4a4fcb550b9a5174fd92a3a2db63c0eaaac/packages/components/src/combobox-control/styles.ts))

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (mirka) [inline packages/components/src/range-control/styles/range-control-styles.ts]: Would it be better/possible to use the `__unstableInputWidth` prop on NumberControl rather than a CSS override?

**Review by mirka — APPROVED** [review]
I noticed some minor non-blocking things, but otherwise looks good to go. Thanks for taking over, Chad!

**Comment by richtabor** [comment]
Looking good!

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [reply packages/components/src/range-control/styles/range-control-styles.ts]: Apologies, I had actually suggested Chad to write code this way!

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (ciampo) [reply packages/components/src/range-control/styles/range-control-styles.ts]: My bad on this one as well! This is also the result of a change that [I suggested](https://github.com/WordPress/gutenberg/pull/49105#discussion_r1221239876) to clean up styles and types a bit, `small` is there to the `sizes` object has, as keys, all possible values that of `size` prop from `NumberControl`.

Feel free to make changes as needed :)

Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: Good point. That only got added when we [decided](https://github.com/WordPress/gutenberg/pull/49105#discussion_r1221239876) to re-use the `NumberControl` prop type instead of typing this bit locally. Can remove the `small` case, but I had to get a little creative with the typing as a result. 24190be736

_Edit: this might be a moot point [if we can scrip this logic altogether](https://github.com/WordPress/gutenberg/pull/49105#discussion_r1235316880)_

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: Either works for me, happy to switch it back to the previous approach that matches your recommendation @mirka. da32f2d92d

File: packages/components/src/range-control/styles/range-control-styles.ts
Comment (chad1008) [reply packages/components/src/range-control/styles/range-control-styles.ts]: That seems to work well for me when testing in Storybook. Was 6fec45f139 along the lines you were thinking? Just to make sure I've implemented it properly.

If so, it has the added benefit of simplifying all of that `inputWidth` logic and typing!

**Comment by chad1008** [comment]
Thanks @mirka and @richtabor!

Lena, I know you dropped a green checkmark after your last review but I've made a few changes based on your feedback, so I'll hold up for a final 👍 and then I'll update the changelog and get this merged.

**Review by mirka — APPROVED** [review]
Looks good, thank you!

It's a bit weird that `rangeHeightValue = 30` is still always used for some margin-top calculations, and ideally that should be refactored to make sense, but I don't consider that a blocker for this PR. Especially since a new RangeControl component is in the pipeline.

**Comment by ciampo** [comment]
This PR likely needs a rebase/merge from trunk to get e2e to pass

**Comment by ciampo** [comment]
For the dev note, see https://github.com/WordPress/gutenberg/issues/46741

**Pull request merged**
